### ISSUE TITLE
How to Access Full 10Hz Camera Sequences in E2E Driving Dataset?

### ISSUE THREAD

**Issue by hayoung-kim** [waymo-research/waymo-open-dataset#918]
Hi,

According to the [camera data section](https://waymo.com/open/data/e2e#camera-data) of the Waymo Open Dataset site, it says that 10Hz camera video sequences are provided for each driving segment:

> This dataset includes images from eight cameras, each capturing a different direction: front, front left, front right, side left, side right, rear, rear left, and rear right. For each direction, a single JPEG image is provided. Alongside the image data, we supply camera intrinsics and extrinsics, which define the camera’s internal parameters and its position relative to the vehicle’s center, respectively. These parameters enable the projection of 3D trajectories onto the camera images. **Each driving segment features 10Hz camera video sequences, with training data spanning 20 seconds and testing data spanning 12 seconds.**

However, in the provided tutorial notebook ([link](https://github.com/waymo-research/waymo-open-dataset/blob/d16af5cf112a2498d659f87e614ad19f20ca2f56/tutorial/tutorial_vision_based_e2e_driving.ipynb)), it seems we can only access the current frame’s images. In `data.frame.images`, there are only 8 camera images corresponding to the current timestamp.

It would be really helpful if there were guidance on how to access the full video sequence, especially how to retrieve camera images from previous frames in a segment.

Thanks in advance for your support!

**Comment by synsin0** [comment]
Does the e2e driving dataset contain labels of objects and maps? Or does it only consist of trajectories and images?

**Comment by synsin0** [comment]
@hayoung-kim You can get the sequence name and the sample index by: 
sequence_name, sample_idx = data.frame.context.name.split('-'). 
Then you can get a continuous sequence of eight cameras in order.

**Comment by sage31** [comment]
I'm also wondering about this, and the rater feedback labels for the training set (they don't seem to be encompassed in the proto anywhere)

**Comment by hayoung-kim** [comment]
@synsin0 Thanks for the info! Sorry to bother you again, but if you happen to know how to actually access the image data using the `sequence_name` and `sample_idx`, I’d really appreciate your help.

I’m trying to retrieve the corresponding camera images for a given frame using those values, but haven’t been able to find a clear way to do it.

For reference, I’m currently parsing the data like this:
```
dataset = tf.data.TFRecordDataset(filenames, compression_type='')
dataset_iter = dataset.as_numpy_iterator()

bytes_example = next(dataset_iter)
data = wod_e2ed_pb2.E2EDFrame()
data.ParseFromString(bytes_example)
```

**Comment by hayoung-kim** [comment]
> Does the e2e driving dataset contain labels of objects and maps? Or does it only consist of trajectories and images?

From what I’ve seen, it just includes trajectories and images.

**Comment by sage31** [comment]
> [@synsin0](https://github.com/synsin0) Thanks for the info! Sorry to bother you again, but if you happen to know how to actually access the image data using the `sequence_name` and `sample_idx`, I’d really appreciate your help.

You could try something like this, which aggregates the frames by seq num:
```
frame_bytes = next(dataset_iter)
data = wod_e2ed_pb2.E2EDFrame()
data.ParseFromString(frame_bytes)
target_sequence_num, sample_idx = data.frame.context.name.split("-")

matching_frames = []
for other_frames in dataset:
    other_data = wod_e2ed_pb2.E2EDFrame()
    other_data.ParseFromString(other_frames.numpy())
    sequence_num, sample = other_data.frame.context.name.split("-")

    if sequence_num == target_sequence_num:
        matching_frames.append(other_data)

# do something with matching_frames, probably sort by frame idx
```

My issue is that I do not have the dataset downloaded (accessing directly from GCS), so the aggregation is unbearably slow--I think the tfrecords must not be sorted by seq num

**Comment by hayoung-kim** [comment]
@synsin0 @sage31 Thank you all for your kind explanations. I was wondering if there was an alternative to this rather brutal method of creating it directly, but it seems there isn’t.

**Comment by synsin0** [comment]
@hayoung-kim @sage31 I find that I cannot find the relative 6-DoF vehicle pose (translation and yaw) during a continuous driving clip. Do they provide this info in e2e data? The vehicle pose (they say ego2global) is always np.eye(4) as I see.

**Comment by hubertsgithub** [comment]
Hi all, our recommendation is to aggregate frames by sequence name and index (https://github.com/waymo-research/waymo-open-dataset/issues/918#issuecomment-2787933116 and https://github.com/waymo-research/waymo-open-dataset/issues/918#issuecomment-2788265949)

Regarding relative pose (translation): these can be inferred from the future trajectory positions., e.g. frame_t.future_trajectory will give you the relative translation for frames t+1,...,t+N.

Thanks for raising this, we'll look into alternatives for streamlining the process.
